main/adv_table.c

```diff
@@ -0,0 +1,174 @@
+/**
+ * @file adv_table.c
+ * @author TheSomeMan
+ * @date 2021-01-19
+ * @copyright Ruuvi Innovations Ltd, license BSD-3-Clause.
+ */
+
+#include "adv_table.h"
+#include <string.h>
+#include <limits.h>
+#include "os_mutex.h"
+
+#define ADV_TABLE_HASH_SIZE (101)
+
+typedef struct adv_report_hash_list_elem_t adv_report_hash_list_elem_t;
+struct adv_report_hash_list_elem_t
+{
+    adv_report_hash_list_elem_t *p_next;
+    adv_report_t *               p_adv_report;
+};
+
+typedef struct adv_hash_table_t
+{
+    adv_report_hash_list_elem_t *p_list[ADV_TABLE_HASH_SIZE];
+} adv_hash_table_t;
+
+static adv_report_table_t          g_adv_reports;
+static os_mutex_t                  gp_adv_reports_mutex;
+static os_mutex_static_t           g_adv_reports_mutex_mem;
+static adv_hash_table_t            g_adv_hash_table;
+static adv_report_hash_list_elem_t g_adv_hash_arr_of_elems[MAX_ADVS_TABLE];
+
+void
+adv_table_init(void)
+{
+    gp_adv_reports_mutex      = os_mutex_create_static(&g_adv_reports_mutex_mem);
+    g_adv_reports.num_of_advs = 0;
+    memset(&g_adv_hash_table, 0, sizeof(g_adv_hash_table));
+    memset(g_adv_hash_arr_of_elems, 0, sizeof(g_adv_hash_arr_of_elems));
+}
+
+void
+adv_table_deinit(void)
+{
+    os_mutex_delete(&gp_adv_reports_mutex);
+}
+
+static bool
+mac_address_is_equal(const mac_address_bin_t *const p_mac1, const mac_address_bin_t *const p_mac2)
+{
+    if (0 == memcmp(p_mac1->mac, p_mac2->mac, MAC_ADDRESS_NUM_BYTES))
+    {
+        return true;
+    }
+    return false;
+}
+
+ADV_TABLE_STATIC
+uint32_t
+adv_report_calc_hash(const mac_address_bin_t *const p_mac)
+{
+    uint32_t     hash_val           = 0;
+    const size_t mac_addr_half_size = sizeof(p_mac->mac) / sizeof(p_mac->mac[0]) / 2;
+    for (uint32_t i = 0; i < mac_addr_half_size; ++i)
+    {
+        hash_val |= ((uint32_t)p_mac->mac[i] ^ (uint32_t)p_mac->mac[i + mac_addr_half_size]) << (i * CHAR_BIT);
```

Comment:
Is there some source which explains this hash function?

Reply:
It's just XOR-based hash calculation - one of the simplest and effective method for hashing.
For example, here is comparison of various hashing functions (XOR, checksum, CRC): https://www.cs.wustl.edu/~jain/papers/ftp/hash_iee.pdf
If this code 24-bit has value is calculated as XOR between two 24-bit parts of MAC address.
Since the higher 24-bits of bluetooth MAC address are fixed for a vendor, so in most practical cases (for devices of one vendor) all hashes  will have unique values.
This 24-bit hash value is divided to some prime number (101 in this case) to calculate an index in the hash table.